src/utils/auth.go

```diff
@@ -25,49 +24,48 @@ func Authenticate(pc **gophercloud.ProviderClient, service string, log logrus.Fi
 	}
 
 	var err error
-	var authOpts gophercloud.AuthOptions
+	var authOpts *clientconfig.AuthInfo
+	var clientOpts clientconfig.ClientOpts
 
 	if _, ok := os.LookupEnv("OS_SWIFT_AUTH_URL"); ok && service == "swift" {
-		log.Infof("Authenticating against Swift using environment variables")
-		authOpts = gophercloud.AuthOptions{
-			IdentityEndpoint: os.Getenv("OS_SWIFT_AUTH_URL"),
-			Username:         os.Getenv("OS_SWIFT_USERNAME"),
-			UserID:           os.Getenv("OS_SWIFT_USER_ID"),
-			Password:         os.Getenv("OS_SWIFT_PASSWORD"),
-			Passcode:         os.Getenv("OS_SWIFT_PASSCODE"),
-			DomainID:         os.Getenv("OS_SWIFT_DOMAIN_ID"),
-			DomainName:       os.Getenv("OS_SWIFT_DOMAIN_NAME"),
-			TenantID:         os.Getenv("OS_SWIFT_TENANT_ID"),
-			TenantName:       os.Getenv("OS_SWIFT_TENANT_NAME"),
+		log.Infof("Authenticating against Swift using special swift environment variables (see README.md)")
+		authOpts = &clientconfig.AuthInfo{
+			ApplicationCredentialID:     os.Getenv("OS_SWIFT_APPLICATION_CREDENTIAL_ID"),
+			ApplicationCredentialName:   os.Getenv("OS_SWIFT_APPLICATION_CREDENTIAL_NAME"),
+			ApplicationCredentialSecret: os.Getenv("OS_SWIFT_APPLICATION_CREDENTIAL_SECRET"),
+			AuthURL:                     os.Getenv("OS_SWIFT_AUTH_URL"),
+			Username:                    os.Getenv("OS_SWIFT_USERNAME"),
+			UserID:                      os.Getenv("OS_SWIFT_USER_ID"),
+			Password:                    os.Getenv("OS_SWIFT_PASSWORD"),
+			DomainID:                    os.Getenv("OS_SWIFT_DOMAIN_ID"),
+			DomainName:                  os.Getenv("OS_SWIFT_DOMAIN_NAME"),
+			ProjectName:                 os.Getenv("OS_SWIFT_PROJECT_NAME"),
+			ProjectID:                   os.Getenv("OS_SWIFT_PROJECT_ID"),
+			UserDomainName:              os.Getenv("OS_SWIFT_USER_DOMAIN_NAME"),
+			UserDomainID:                os.Getenv("OS_SWIFT_USER_DOMAIN_ID"),
+			ProjectDomainName:           os.Getenv("OS_SWIFT_PROJECT_DOMAIN_NAME"),
+			ProjectDomainID:             os.Getenv("OS_SWIFT_PROJECT_DOMAIN_ID"),
 		}
+		clientOpts.AuthInfo = authOpts
 	} else {
-		log.Infof("Authenticating against Openstack using environment variables")
-		authOpts, err = openstack.AuthOptionsFromEnv()
-		if err != nil {
-			return err
-		}
+		log.Infof("Trying to authenticate against Openstack using environment variables (including application credentials) or using files ~/.config/openstack/clouds.yaml, /etc/openstack/clouds.yaml and ./clouds.yaml")
 	}
 
-	authOpts.AllowReauth = true
+	allowReauth := true
+	authOpts.AllowReauth = &allowReauth
```

Comment:
I guess the authOpts might be null if no 'OS_SWIFT_AUTH_URL' detectd which results in the nil pointer or something else here.